[HLD] Readonly appearance for disabled controls (#2592)
ni/nimble: specs/angular-readonly-form-styling/README.md

> # Angular read-only form styling

Reviewer: Is anything in the proposal Angular-specific anymore? If not, could you update the document and PR titles to be more generic?

With some changes to nimble's controls, applications could much more easily create forms with read-only controls rather than disabled controls.

An example can be seen below of a page that currently has disabled controls and the ideal UI.

Current UI with disabled controls:

![disabled form](disabled-form.png)

Improved UI with read-only controls:
![readonly form](readonly-form.png)

Note that in the improved UI screenshot, there are likely still additional improvements that should be made. These improvements include, but are not limited to, updating the `nimble-switch` and `nimble-checkbox` instances to no longer be disabled. However, a decision has not yet been made regarding how these components should be represented when read-only.

## Links To Relevant Work Items and Reference Material

[SLE control pattern docs](https://stratus-storybook.ni.dev/?path=/docs/patterns-control-text--docs)

## Implementation / Design

To enable this new design, the following changes need to be made within the nimble repo, each of which is discussed in more detail below:

> - Introduce `readonly-disabled-appearance` configuration on the `nimble-text-field`, `nimble-number-field`, `nimble-text-area`, `nimble-select`, and `nimble-combobox`

Reviewer 2: Think I'd prefer an `appearance-` prefix to align with `appearance` / `appearance-variant`.
I think just keep the name describing the state (it's a readonly visual that today only does anything when the control is in the disabled state, but I don't think that part needs to be encoded in the name).

So I'd go for `appearance-readonly` with the docs describing that it only applies when disabled.

Alternatives:
- `appearance-readonly` (my pref)
- `appearance-force-readonly`
- `appearance-override-readonly`
- `appearance-show-readonly`
- `appearance-as-readonly`
- `appearance-readonly-override`
- `appearance-readonly-force`

### Frameless select, combobox, and number field

The visual design for read-only controls in forms is for them to be rendered without any borders or backgrounds. This matches the `frameless` appearance nimble has today for the `nimble-text-field`. Therefore, the select, combobox, and number field will all be updated to also support a `frameless` appearance.

### Full-bleed select, combobox, and number field

The visual design for read-only controls in forms is for them to be rendered without any left padding on the value (i.e. the value is aligned with the label). This matches the `full-bleed` configuration nimble has today for the `nimble-text-field`. Therefore, the select, combobox, and number field will all be updated to also support being `full-bleed`.

Similar to the `nimble-text-field`, these components will only support `full-bleed` when their appearance is `frameless`.

## Separation of responsibility

With the changes described above implemented in nimble, an application can do the following:

- Create an Angular form with nimble controls configured with `readOnlyDisabledAppearance`

> - Update controls to use `frameless` appearance and `full-bleed` when they are disabled and will be styled as read-only

Author: @jattasNI and @rajsite - I'm curious if you think this is the correct separation of responsibility with the new direction to leave form controls as disabled but style them as readonly based on the presence of a new attribute. In that scenario, should nimble be more opinionated and style things as `frameless` + `full-bleed` regardless how these properties are configured on the component?

Reviewer 2: I think conceptually we want `readonly` and `disabled[appearance-readonly]` to be as close as possible visually. Right now [`readonly` seems to respond to different appearance modes](https://nimble.ni.dev/storybook/index.html?path=/story/components-text-field--block-text-field&args=readonly:!true).

If we were to adopt the behavior described for `appearance-readonly` we should change it for `readonly` as well. Not necessarily a deal breaker just thinking out loud. I don't think I feel strongly one way or another (struggling to think off pros/cons of more flexibility on readonly states looking "more control like" vs strictly "more text like").

Author: I don't think we want to change the way `readonly` works. You can see some screenshots in the "Alternatives" section of the document of where we use `readonly` controls with non-frameless appearances. We shouldn't break those usages.

If we want to keep `appearance-readonly` aligned with `readonly` (which I think is a reasonable desire), then I don't think we should "force" `frameless` + `full-bleed` styles with `appearance-readonly`.

## Implementation / Design

To enable this new design, the following changes need to be made within the nimble repo, each of which is discussed in more detail below:

- Introduce `readonly` configuration on the `nimble-select` and `nimble-combobox`
- Introduce `frameless` appearance to the `nimble-select`, `nimble-combobox`, and `nimble-number-field`
- Introduce `full-bleed` configuration on the `nimble-select`, `nimble-combobox`, and `nimble-number-field`
- Introduce a `readOnlyWhenDisabled` configuration on the control value accessors in Angular for the `nimble-select`, `nimble-combobox`, `nimble-text-area`, `nimble-text-field`, and `nimble-number-field`

### Readonly select and combobox

The `nimble-select` and `nimble-combobox` will be updated to have a new boolean attribute named `readonly`, which will be associated with the `readOnly` boolean property.

When these components are readonly, they will have the following characteristics:

- Tab focusable
- Cannot change their value using the mouse or keyboard
- Cannot open their popup
- Drop-down arrow (select) and drop-down button (combobox) will be hidden

> - Aria updates when readonly:

Reviewer: Looks like you missed the first half of this comment (or I missed where you addressed it).

## Separation of responsibility

With the changes described above implemented in nimble, an application can do the following:

- Create an Angular form with nimble controls configured with `readonlyWhenDisabled`
- Update controls to use `frameless` appearance and `full-bleed` when the are readonly
    - Note: This can be done in a single-source way if a client creates an Angular directive that applies the desired attributes on nimble components based on the state of the `readonly` attribute on that component.

This allows nimble to only make generic changes. Nimble will be responsible for:
- Styling the component correctly given the `readonly` state, appearance, and `full-bleed` state
- Making the component `readonly` when the form control is disabled and `readonly-when-disabled` is configured on the control value accessor directive

> ## Alternative Implementations / Designs

Author: I'll add this as an alternative, but we do have `readonly` text fields that aren't `full-bleed` or `frameless` today. Therefore, I don't think we want to couple these concepts together.

> ## Alternative Implementations / Designs

Reviewer: Can you share an example? I want to understand if they're legitimate use cases or just clients that didn't realize they should be setting `full-bleed` and `frameless` alongside `readonly`.

Maybe a good outcome of this would be for us to enhance our usage guidance around these attributes. The text field storybook doesn't mention much beyond how the attributes will behave.

> ## Alternative Implementations / Designs

Author: Yes, my plan is to include some screenshots of where it is used. I believe they are legitimate use cases.

- Introduce `readonly` configuration on the `nimble-select` and `nimble-combobox`
- Introduce `frameless` appearance to the `nimble-select`, `nimble-combobox`, and `nimble-number-field`
- Introduce `full-bleed` configuration on the `nimble-select`, `nimble-combobox`, and `nimble-number-field`
- Introduce a `readOnlyWhenDisabled` configuration on the control value accessors in Angular for the `nimble-select`, `nimble-combobox`, `nimble-text-area`, `nimble-text-field`, and `nimble-number-field`

### Read-only select and combobox

The `nimble-select` and `nimble-combobox` will be updated to have a new boolean attribute named `readonly`, which will be associated with the `readOnly` boolean property.

When these components are read-only, they will have the following characteristics:

- Tab focusable
- Cannot change their value using the mouse or keyboard
- Cannot open their popup
- Drop-down arrow (select) and drop-down button (combobox) will be hidden
- ARIA updates when read-only:
    - `aria-haspopup` set to `null`
    - `aria-readonly` set to `"true"`

### Frameless select, combobox, and number field

> The visual design for rea-donly controls in forms is for them to be rendered without any borders or backgrounds. This matches the `frameless` appearance nimble has today for the `nimble-text-field`. Therefore, the select, combobox, and number field will all be updated to also support a `frameless` appearance.

Reviewer: ```suggestion
The visual design for read-only controls in forms is for them to be rendered without any borders or backgrounds. This matches the `frameless` appearance nimble has today for the `nimble-text-field`. Therefore, the select, combobox, and number field will all be updated to also support a `frameless` appearance.
```

## Separation of responsibility

With the changes described above implemented in nimble, an application can do the following:

- Create an Angular form with nimble controls configured with `readonlyWhenDisabled`
- Update controls to use `frameless` appearance and `full-bleed` when they are read-only
    - Note: This can be done in a single-source way if a client creates an Angular directive that applies the desired attributes on nimble components based on the state of the `readonly` attribute on that component.

This allows nimble to only make unopinionated changes. Nimble will be responsible for:
- Styling the component correctly given the `readonly` state, appearance, and `full-bleed` state
- Making the component `readonly` when the form control is disabled and `readonly-when-disabled` is configured on the control value accessor directive

## Angular updates

> In additional to the changes to the Angular control value accessors described above, all new `nimble-components` attributes will be exposed in the Angular directives.

Reviewer: ```suggestion
In addition to the changes to the Angular control value accessors described above, all new `nimble-components` attributes will be exposed in the Angular directives.
```

![readonly text field example 1](readonly-text-field-1.png)

![readonly text field example 2](readonly-text-field-2.png)

**Repurpose the `disabled` state**

Rather than adding a `readonly` state on the select and combobox, we could re-style these components to look read-only when they are disabled. However, this approach has multiple problems including:

- Native inputs support both `readonly` and `disabled`, and there are important differences between them. For example, read-only controls are focusable and disabled controls are not.
- There are still use cases for disabled controls, and conflating the meaning of "disabled" and "read-only" would cause us difficulty in supporting both.

> ## Open Issues

Reviewer: What's the plan to resolve the open issues? Just need feedback from Fred/Brandon on requirements? Or are there technical challenges?

I'm ok if we resolve them in a follow up PR.

Author: I don't think there are technical challenges, particularly around adding different appearance modes to the text-area.

I think this is mostly a question for Fred/Brandon, but I thought it made sense to get general consensus on this approach before extending it too far to other components.

# Angular read-only form styling

## Problem Statement

There is a desire to render Angular forms with read-only controls rather than disabled controls when the control in the Angular form is marked as disabled. This is because there are use cases where the configuration represented by nimble controls is not editable but is still relevant and should be easily readable. Additionally, the only way to mark something in an Angular form as "not editable" is to mark it as disabled (i.e. there is no concept of "read only" in an Angular form). However, there is quite a bit of logic built into Angular forms and Angular's control value accessors around what it means to render a component bound to a disabled form control.

With some changes to nimble's controls and nimble's Angular control value accessors, then Angular applications could much more easily create forms with read-only controls rather than disabled controls.

An example can be seen below of a page that currently has disabled controls and the ideal UI.

Current UI with disabled controls:

> ![disabled form](disabled-form.png)

Reviewer 3: ```suggestion

![disabled form](disabled-form.png)
```
In my browser, it's putting the image directly to the right of the text. Perhaps a newline fixes that?

When these components are read-only, they will have the following characteristics:

- Tab focusable
- Cannot change their value using the mouse or keyboard
- Cannot open their popup
- Drop-down arrow (select) and drop-down button (combobox) will be hidden
- ARIA updates when read-only:
    - `aria-haspopup` set to `null`
    - `aria-readonly` set to `"true"`
    - Note: While the native select does not support `readonly`, the `combobox` role does support `aria-readonly`, so that can be used in combination with the existing role of both the `nimble-select` and `nimble-combobox`. With these two changes locally, storybook does not report any new accessibility violations.

### Frameless select, combobox, and number field

The visual design for read-only controls in forms is for them to be rendered without any borders or backgrounds. This matches the `frameless` appearance nimble has today for the `nimble-text-field`. Therefore, the select, combobox, and number field will all be updated to also support a `frameless` appearance.

### Full-bleed select, combobox, and number field

> The visual design for read-only controls in forms is for them to be rendered without any left padding on the value (i.e. the value is aligned with the label). This matches the `full-bleed` configuration nimble has today for the `nimble-text-field`. Therefore, the select, combobox, and number field will all be updated to also support being `full-bleed`.

Reviewer 3: Is it worth noting that, like the `nimble-text-field`, the `full-bleed` configuration only applies to the `frameless` appearance?

Author: Yes, that's a great point. My plan is to have `full-bleed` + `frameless` to work together the same way as it does on the `nimble-text-field` today. I should explicitly call that out here.

With some changes to nimble's controls and nimble's Angular control value accessors, then Angular applications could much more easily create forms with read-only controls rather than disabled controls.

An example can be seen below of a page that currently has disabled controls and the ideal UI.

Current UI with disabled controls:

![disabled form](disabled-form.png)

Improved UI with read-only controls:
![readonly form](readonly-form.png)

Note that in the improved UI screenshot, there are likely still additional improvements that should be made. These improvements include, but are not limited to, updating the `nimble-switch` and `nimble-checkbox` instances to no longer be disabled. However, a decision has not yet been made regarding how these components should be represented when read-only.

## Links To Relevant Work Items and Reference Material

[SLE control pattern docs](https://stratus-storybook.ni.dev/?path=/docs/patterns-control-text--docs)

> ## Implementation / Design

Author: Today, the disabled state generally wins out. However, I didn't do a thorough audit of the controls, so I'm not 100% sure that `disabled` is styled identically as `readonly + disabled` for all cases.

This brings up an interesting point. Another alternative to consider is that if we had `readonly` styling win out over `disabled` styling, then we wouldn't need to change directive or theme-provider behavior. Instead, we could do something like this in Angular:

```html
<nimble-text-field
    formControlName="myControl"
    [readonly]="form.controls.myControl.disabled"
    [appearance]="form.controls.myControl.disabled ? 'frameless' : 'underline'"
    [full-bleed]="form.controls.myControl.disabled"
>
    Control label
</nimble-text-field>
```

> ## Implementation / Design

Reviewer 2: Interesting concept though I think it feels a bit weird for disabled to not override everything else. And then I feel like I could imagine both modes being useful:

- disabled form behavior + readonly interaction
- readonly form behavior + disabled interaction

Searching around there is a whatwg issue around these discussions: https://github.com/whatwg/html/issues/2311

The issue reminded me about another point that only a [subset of form controls support readonly](https://developer.mozilla.org/en-US/docs/Web/HTML/Reference/Attributes/readonly#overview) today.

What about having separate attributes that are only used in conjunction?

- disabled form behavior + readonly interaction `[disabled][readonly-visible]`
- readonly form behavior + disabled interaction `[readonly][disabled-visible]`

With that I think the `readonly-visible` and `disabled-visible` would only have a behavior when combined with their corresponding state. Naming ideas:

-  `[disabled][readonly-visible]`
-  `[disabled][readonly-interaction]` (might actually be leaning to this)
-  `[disabled][force-readonly-visible]`
-  `[disabled][force-readonly-interaction]`
-  `[disabled][force-readonly]`
-  `[disabled][disabled-variant]`
-  `[disabled][disabled-variant="readonly"]`

With something like that we can avoid defining new semantics / behaviors for `[readonly][diabled]`, we and aren't constrained to a specific definition of that combo longer-term, and we don't have to introduce new definitions of readonly that aren't present today in DOM (i.e. readonly select, etc).

> ## Implementation / Design

Author: > I think it feels a bit weird for disabled to not override everything else

I agree with that, which is why I hadn't originally considered this a valid option.

> What about having separate attributes that are only used in conjunction?

I wouldn't be opposed to doing something like `[disabled][readonly-interaction]`. Is your thought that `[readonly-interaction]` would have no effect if the control was not disabled?

> ## Implementation / Design

Reviewer 2: > Is your thought that [readonly-interaction] would have no effect if the control was not disabled?

Yep! It's like the mental model of appearance-variant applying to only some modes of appearance.

> ## Implementation / Design

Author: I've updated the document to reflect this new direction. Given that we are going to keep the control as disabled from an ARIA and focusability point of view, I didn't think `readonly-interaction` was an appropriate term. Therefore, I've proposed `readonly-disabled-appearance`. I'm not sold on that name, either, so I'm open to counter proposals.

### Control value accessor updates

Part of Angular's `ControlValueAccessor` interface is an optional `setDisabledState` function, which is [documented here](https://angular.dev/api/forms/ControlValueAccessor#setDisabledState). All of the control value accessors currently implemented in nimble rely on an implementation of `setDisabledState` that looks like this:

```ts
setDisabledState(isDisabled: boolean): void {
    this.setProperty('disabled', isDisabled);
}
```

Because of the way that Angular injects an `NG_VALUE_ACCESSOR` for a control and because there should only be one `NG_VALUE_ACCESSOR` for a given control, the only way to change the behavior of a disabled form control is to change the control value accessor associated with it to have a different implementation of `setDisabledState`. Therefore, nimble's control value accessor directives will be updated to have the option to set `readOnly` on a control rather than `disabled`. This implementation will look like this:

> ```ts

Author: Closing as this is no longer relevant with the new direction.

@Input('readonly-when-disabled') public readOnlyWhenDisabled: boolean;

public override setDisabledState(isDisabled: boolean): void {
    if (this.readOnlyWhenDisabled) {
        this.setProperty('readOnly', isDisabled);
    } else {
        super.setDisabledState(isDisabled);
    }
}
```

The `readOnlyWhenDisabled` input will default to `false`, so existing clients of the directives will not be affected by these changes.

> Additionally, nimble's control value accessors will need a small amount of logic to handle `readonly` being bound directly in a template that is using `ngModel`. This logic will be adapted from the [Angular code](https://github.com/angular/angular/blob/1b8e7ab9fe46901979389b377be4232e11092260/packages/forms/src/directives/ng_model.ts#L269) that works the same way for the `disabled` state on a control.

Author: Closing as this is no longer relevant with the new direction.

## Separation of responsibility

With the changes described above implemented in nimble, an application can do the following:

> - Create an Angular form with nimble controls configured with `readonlyWhenDisabled`

Author: Closing as this is no longer relevant to the new direction.

- Update controls to use `frameless` appearance and `full-bleed` when they are read-only
    - Note: This can be done in a single-source way if a client creates an Angular directive that applies the desired attributes on nimble components based on the state of the `readonly` attribute on that component.

This allows nimble to only make unopinionated changes. Nimble will be responsible for:
- Styling the component correctly given the `readonly` state, appearance, and `full-bleed` state
- Making the component `readonly` when the form control is disabled and `readonly-when-disabled` is configured on the control value accessor directive

## Angular updates

In addition to the changes to the Angular control value accessors described above, all new `nimble-components` attributes will be exposed in the Angular directives.

## Blazor updates

Initially it is out of scope to add equivalent integration for Blazor forms. All new `nimble-components` attributes will be exposed in NimbleBlazor.

> ## Alternative Implementations / Designs

Reviewer: With this proposal would we get rid of the `readonly` attributes in `nimble-components`? So that you only ever set the `disabled` attribute (and Nimble decides how to interpret that attribute's impact on styling and focus behavior based on `readonly-when-disabled`)? I'm guessing yes, since otherwise we'd end up with redundant ways of setting readonly, which would mean possible conflicts/confusion when they're set differently.

If I'm understanding correctly I think an important distinction between the original proposal and this one is:
1. the original proposal means "treat the **form model** disabled property to mean that the nimble component is readonly"
2. this proposal means "treat the **nimble component** disabled attribute to mean that the nimble component is readonly"

As an initial reaction I think I like 1 better. To me it feels like the lack of a readonly property on the form model is a limitation of Angular which shouldn't necessarily leak into how we design our component APIs. And if we're designing our component APIs in a vacuum I like the original proposal because it keeps the API pretty consistent with what we did for the text field.

But I'm interested in other opinions or being told that I'm misunderstanding something about this proposal.

> ## Alternative Implementations / Designs

Author: This seems like a promising direction to me. I'll plan to do a little bit of prototyping to see how it works in practice.

I do want to clarify my understanding of the proposal -- when `readonly-when-disabled` is set on the theme provider, `disabled` and `readonly` on the control would have identical impacts, correct? Meaning `readonly`, `disabled`, and `readonly + disabled` would all act/look like readonly controls with respect to ARIA, visuals, and tab index, correct?

> ## Alternative Implementations / Designs

Reviewer: Could you capture the theme provider approach as an alternative? I could see some benefit to specifying this behavior for an entire application so it might be useful to record why we decided not to.
